src/Mvc/Mvc.Core/src/Formatters/TranscodingReadStream.cs

```diff
@@ -90,76 +100,50 @@ public override async Task<int> ReadAsync(byte[] buffer, int offset, int count,
-                    // is insufficient if the character requires more than 2 bytes to represent. In this case, read
-                    // content in to an overflow buffer and fill up the passed in buffer.
-                    _encoder.Convert(
-                        _charBuffer,
-                        _overflowBuffer.Array,
-                        flush: false,
-                        out var charsUsed,
-                        out var bytesUsed,
-                        out _);
+                case OperationStatus.Done:
+                    return bytesWritten;
+
+                case OperationStatus.DestinationTooSmall:
+                    if (bytesWritten != 0)
+                    {
+                        return bytesWritten;
+                    }
 
-                    _charBuffer = _charBuffer.Slice(charsUsed);
+                    // Overflow buffer is always empty when we get here and we can use it's full length to write contents to.
+                    Utf8.FromUtf16(_charBuffer, _overflowBuffer.Array, out var overFlowChars, out var overflowBytes, isFinalBlock: false);
+                    Debug.Assert(overflowBytes > 0 && overFlowChars > 0, "We expect writes to the overflow buffer to always succeed since it is large enough to accomodate at least one char.");
+                    Debug.Assert(readBuffer.Count < overflowBytes);
 
-                    Debug.Assert(readBuffer.Count < bytesUsed);
+                    _charBuffer = _charBuffer.Slice(overFlowChars);
                     _overflowBuffer.Array.AsSpan(0, readBuffer.Count).CopyTo(readBuffer);
 
                     _overflowBuffer = new ArraySegment<byte>(
                         _overflowBuffer.Array,
                         readBuffer.Count,
-                        bytesUsed - readBuffer.Count);
-
-                    totalBytes += readBuffer.Count;
-                    // At this point we're done writing.
-                    break;
-                }
-                else
-                {
-                    _encoder.Convert(
-                        _charBuffer,
-                        readBuffer,
-                        flush: allContentRead,
-                        out var charsUsed,
-                        out bytesEncoded,
-                        out encoderCompleted);
-
-                    totalBytes += bytesEncoded;
-                    _charBuffer = _charBuffer.Slice(charsUsed);
-                    readBuffer = readBuffer.Slice(bytesEncoded);
-                }
-
-                // We need to exit in one of the 2 conditions:
-                // * encoderCompleted will return false if "buffer" was too small for all the chars to be encoded.
-                // * no bytes were converted in an iteration. This can occur if there wasn't any input.
-            } while (encoderCompleted && bytesEncoded > 0);
-
-            return totalBytes;
+                        overflowBytes - readBuffer.Count);
```

Comment:
What guarantees do we have about overflowBytes relative to the size of the overflowBuffer?

Reply:
Where does the overflowBuffer get reset?

Reply:
should `ReadInputChars` reset it?

Reply:
I guess this is where?

Reply:
`overflowBytes` represents the number of bytes written to in the buffer. It's guaranteed to be smaller than the input array size. We reset it on Line 131, essentially any time we figure out there isn't enough content in the read buffer. We only ever get to `ReadInputChars` if it's been emptied, so I'm not sure if it helps reseting it there.